ui/src/app.tsx

```diff
@@ -1,43 +1,93 @@
-import React, { useState } from "react";
-import { HashRouter, Route, Switch } from "react-router-dom";
+import * as _ from "lodash";
+import React, { useCallback, useState } from "react";
+import {
+  HashRouter,
+  Redirect,
+  Route,
+  Switch,
+  useParams,
+  useRouteMatch,
+} from "react-router-dom";
 import "./app.css";
 import { Cohort } from "./cohort";
 import { CohortUpdaterProvider } from "./cohortUpdaterContext";
+import { Datasets } from "./datasets";
 import Edit from "./edit";
 import Overview from "./overview";
+import { UserData } from "./userData";
 
 type AppProps = {
-  underlayName: string;
+  underlayNames: string[];
   entityName: string;
 };
 
 export default function App(props: AppProps) {
-  const [cohort, setCohort] = useState<Cohort>(
-    new Cohort(
-      props.underlayName,
-      props.entityName,
-      // TODO(tjennison): Populate from an actual source.
-      [
-        "person_id",
-        "condition_occurrence_id",
-        "condition_concept_id",
-        "condition_name",
-      ]
-    )
+  const [userData, setUserData] = useState<UserData>(
+    new UserData(props.entityName)
+  );
+
+  const update = useCallback(
+    (callback: (userData: UserData) => void) => {
+      const newUserData = _.cloneDeep(userData);
+      callback(newUserData);
```

Comment:
I see. I think this partially gets back to the usage of classes vs plain objects/interfaces. With the UserData class, the clone is probably necessary. In the AoU codebase if we're setting a plain object we can use approaches like:

```
setUserData({
  ...userData,
  cohorts: []
});
```

Or use the functional programming lodash package which has a side-effect free `.set()` (using a similar splay operator approach).


> React will bail out earlier since the parent objects aren't new objects

Is this true? That's not how React state change detection works in general, though admittedly I haven't tested this in depth with hooks so it's possible there are new caveats that I'm not aware of. Regardless, I wasn't suggesting to reuse the same object.

Reply:
Maybe my understanding is wrong then, since I don't think hooks should change how it works. Are you saying that even if the root object is the same, react will still chase through the entire tree of components below to see if any of the children have changed? I thought that once it reached an object where none of the props or state had changed it would stop.

Reply:
w.r.t. React child rerendering logic - that sounds correct to me. I don't see how component hierarchies are relevant to this though - I thought you were talking about React's logic for change detection when you make a state update, i.e. "what are the side-effects of calling `setUserData()` with different inputs?" (same object, vs new object)

Anyways, this is mostly a theoretical point. I'm not suggesting you reuse the same object; it is not recommended to mutate a state variable outside the normal React interfaces - creating a copy is preferred.

Reply:
Doesn't that mean I need to, at a minimum, duplicate every object from the root to whatever is changing? Otherwise the child won't have a chance to update since one of its parents didn't update? Since there's a single state object and the other components are just working on children of it?

Reply:
Lets chat at next standup, I think it's getting too hypothetical for me to be sure I'm following at this point.